packages/core/src/modules/connections/DidExchangeProtocol.ts

```diff
@@ -0,0 +1,462 @@
+import type { Logger } from '../../logger'
+import type { ConnectionRecord } from './repository'
+import type { Routing } from './services/ConnectionService'
+
+import { convertPublicKeyToX25519 } from '@stablelib/ed25519'
+import { inject, Lifecycle, scoped } from 'tsyringe'
+
+import { AgentConfig } from '../../agent/AgentConfig'
+import { InjectionSymbols } from '../../constants'
+import { KeyType } from '../../crypto'
+import { JwsService } from '../../crypto/JwsService'
+import { Attachment, AttachmentData } from '../../decorators/attachment/Attachment'
+import { AriesFrameworkError } from '../../error'
+import { JsonEncoder } from '../../utils/JsonEncoder'
+import { JsonTransformer } from '../../utils/JsonTransformer'
+import { uuid } from '../../utils/uuid'
+import { Wallet } from '../../wallet/Wallet'
+import { DidCommService, DidDocument, DidDocumentBuilder, Key } from '../dids'
+import { DidDocumentRole } from '../dids/domain/DidDocumentRole'
+import { getEd25519VerificationMethod } from '../dids/domain/key-type/ed25519'
+import { getX25519VerificationMethod } from '../dids/domain/key-type/x25519'
+import { DidKey } from '../dids/methods/key/DidKey'
+import { DidPeer, PeerDidNumAlgo } from '../dids/methods/peer/DidPeer'
+import { DidRecord, DidRepository } from '../dids/repository'
+import { ProblemReportError } from '../problem-reports'
+
+import { DidExchangeStateMachine } from './DidExchangeStateMachine'
+import { DidExchangeCompleteMessage } from './messages/DidExchangeCompleteMessage'
+import { DidExchangeRequestMessage } from './messages/DidExchangeRequestMessage'
+import { DidExchangeResponseMessage } from './messages/DidExchangeResponseMessage'
+import { ConnectionService } from './services'
+
+interface DidExchangeRequestParams {
+  label?: string
+  goal?: string
+  goalCode?: string
+  routing: Routing
+  autoAcceptConnection?: boolean
+}
+
+@scoped(Lifecycle.ContainerScoped)
+export class DidExchangeProtocol {
+  private config: AgentConfig
+  private connectionService: ConnectionService
+  private jwsService: JwsService
+  private logger: Logger
+  private didRepository: DidRepository
+
+  public constructor(
+    @inject(InjectionSymbols.Wallet) wallet: Wallet,
+    config: AgentConfig,
+    connectionService: ConnectionService,
+    didRepository: DidRepository,
+    jwsService: JwsService
+  ) {
+    this.config = config
+    this.connectionService = connectionService
+    this.didRepository = didRepository
+    this.jwsService = jwsService
+    this.logger = config.logger
+  }
+
+  public async createRequest(
+    connectionRecord: ConnectionRecord,
+    params: DidExchangeRequestParams
+  ): Promise<DidExchangeRequestMessage> {
+    this.logger.debug(`Create message ${DidExchangeRequestMessage.type} start`, connectionRecord)
+
+    if (!connectionRecord.invitation) {
+      throw new AriesFrameworkError('Connection invitation is missing.')
+    }
```

Comment:
This implementation is still leaning on the 0160 invitation. It feels limiting, as we don't store the OOB invitation anywhere now even though it is richer than the connection invitation message